### PR TITLE
Add parent domain to imap/smtp check

### LINKED ISSUE #5372: Improve auto configuration for mail accounts
### Feature Request

* [x] Remove connectivity check for port 585. According to Wikipedia port 585 was intended for IMAP over TLS/SSL but is deregistered in favour of port 993.

* [x] Hosting providers may split the services. For example the mx record for a domain test.com might be mx.provider.com. But the imap service is reachable at imap.provider.com. Mail already has a check for [imap.mx.provider.com](https://github.com/nextcloud/mail/blob/6b5e0b9db3d0338a8e7adffe70d8128fa8e93419/lib/Service/AutoConfig/ImapConnectivityTester.php#L75). We should consider to check also the parent domain imap.provider.com.

### PR DESCRIPTION
**Pull request by miaulalala**
Fixes #5372

### CODE REVIEW

**Review by ChristophWurst — COMMENTED** [review]
This looks complicated. Could we have tests? :)

File: lib/Service/AutoConfig/MxRecord.php [hunk 1]
```diff
@@ -56,4 +56,17 @@ public function query(string $host) {
...
+	public function stripSubdomain(string $domain): string {
+		$labels = explode('.', $domain);
+
+		$top = count($labels) >= 2 ? array_pop($labels) : '';
+		$second = array_pop($labels);
+
+		return $second . '.' . $top;
+	}
+
+	public function getAllRecords(array $mxHosts): array {
```
Comment (kesselb) [inline lib/Service/AutoConfig/MxRecord.php:69]: I would say the method is not getting all records but sanitize a list of records. Can we find a better name for it?

File: lib/Service/AutoConfig/MxRecord.php [hunk 2]
```diff
@@ -56,4 +56,17 @@ public function query(string $host) {
 		// TODO: sort by weight
 		return $mxRecords;
 	}
+
+	public function stripSubdomain(string $domain): string {
```
Comment (kesselb) [inline lib/Service/AutoConfig/MxRecord.php:60]: I suggest to make the method private. We can use `$this->invokePrivate($object, 'methodName');` in a test case to execute private methods.

File: lib/Service/AutoConfig/MxRecord.php (hunk 1 above)
Comment (miaulalala) [reply lib/Service/AutoConfig/MxRecord.php:69]: sure, getSanitizedRecords?

File: tests/Unit/Service/Autoconfig/MxRecordTest.php [hunk 3]
```diff
@@ -52,4 +52,29 @@ public function testQueryNoRecord() {
 
 		$this->assertFalse($records);
 	}
+
+	public function testStripSubdomain() {
```
Comment (ChristophWurst) [inline tests/Unit/Service/Autoconfig/MxRecordTest.php:56]:
```suggestion
	public function testStripSubdomain(): void {
```

File: tests/Unit/Service/Autoconfig/MxRecordTest.php [hunk 4]
```diff
@@ -52,4 +52,29 @@ public function testQueryNoRecord() {
...
+		$record = $this->record->stripSubdomain('mx.google.com');
+		$this->assertEquals('google.com', $record);
+
+		$record = $this->record->stripSubdomain('shgfghd.fhdjfhsdf.gggg.google.com');
+		$this->assertEquals('google.com', $record);
+	}
+
+	public function testGetAllRecords() {
```
Comment (ChristophWurst) [inline tests/Unit/Service/Autoconfig/MxRecordTest.php]:
```suggestion
	public function testGetAllRecords(): void {
```

**Review by ChristophWurst — APPROVED** [review]
It aint much, but it's honest typing

**Pull request merged**